[ai-ish] docs: Explain security/data access model for E2EE push notifications. (#39055)
zulip/zulip: docs/production/mobile-push-notifications.md

Google's and Apple's security model for mobile push notifications does not allow
self-hosted Zulip servers to directly send mobile notifications to the Zulip
mobile apps. The Zulip Mobile Push Notification Service solves this problem by
forwarding E2EE mobile push notifications generated by your server to the Zulip

> mobile apps.

Reviewer: This reads as if it's narrowing the scope of what sorts of notifications get forwarded (which wouldn't be true).

Meanwhile if the aim is to reassure people that the data is encrypted, I think this is probably too terse to be effective at that. Perhaps an added sentence at the end? Like:
> With Zulip Server 12.0+ and an up-to-date Zulip mobile app, the notifications are encrypted end-to-end from your server to your mobile device.

Google's and Apple's security model for mobile push notifications does not allow
self-hosted Zulip servers to directly send mobile notifications to the Zulip
mobile apps. The Zulip Mobile Push Notification Service solves this problem by

> forwarding E2EE mobile push notifications generated by your server to the Zulip
> mobile apps. With Zulip Server 12.0+ and an up-to-date Zulip mobile app, the
> notifications are encrypted end-to-end from your server to your mobile device.

Reviewer: ```suggestion
forwarding mobile push notifications generated by your server to the Zulip
mobile apps. With Zulip Server 12.0+ and an up-to-date Zulip mobile app, the
notifications are encrypted end-to-end from your server to your mobile device.
```

(cont'd from https://github.com/zulip/zulip/pull/39055#discussion_r3120663109 )

  - A timestamp.
  - The message's content.

  Zulip 11.0+ has an organization-level setting available to disable
  message content being sent via the push notification bouncer (i.e.,
  message content will be replaced with `New message`), for clients
  that don't support the new end-to-end encrypted notifications
  protocol. As of July 2025, this setting makes push notifications
  significantly less usable, since mobile client support for
  end-to-end encrypted push notifications is not yet available.

  (Prior to Zulip 11.0, this functionality was available via the
  `PUSH_NOTIFICATION_REDACT_CONTENT` server-level setting).

- All of the network requests (both from Zulip servers to the Push
  Notification Service and from the Push Notification Service to the
  relevant Google and Apple services) are encrypted over the wire with
  SSL/TLS.
- The code for the push notification forwarding service is 100% open
  source and available as part of the
  [Zulip server project on GitHub](https://github.com/zulip/zulip)
  (specifically, [here](https://github.com/zulip/zulip/tree/main/zilencer)).
- The push notification forwarding servers are professionally managed

>   by a small team of security-sensitive engineers.

Reviewer: This remains accurate, though I don't know if it's helpful to say it.

Author: @timabbott felt that it's not (in shoulder review), since the team now doesn't have any ability to access data.

Reviewer: Ah, I meant in the section at the bottom about the legacy protocol. (That's where most of the surrounding items are getting moved to anyway.)

Reviewer: ^

If you have any questions about the security model, [contact Zulip

> support](https://zulip.com/help/contact-support).

Reviewer: This definitely seems still helpful to say.

> about recipients, or any other Zulip-specific metadata about the notification.

Reviewer: "information about recipients" is quite broad, and might be taken to include some information that we and Apple and Google do unavoidably see: namely, just enough to route the notifications to the right users' devices, in the form of those APNs or FCM tokens.

Maybe something like "the sender's or recipients' names or avatars"?

Author: I was trying to squash together channel/topic recipients and DMs, but I think it just needs to be broken out.

Author: Copying over a sentence from the help center:

> E2EE ensures that message **content** and **metadata**
(including the sender's name, and the recipient's name or channel and topic
where the message was sent) are not visible to Apple, Google, or Zulip's
[Mobile Push Notification Service](https://zulip.readthedocs.io/en/stable/production/mobile-push-notifications.html).

> For each notification, the Push Notification Service receives and forwards:

Reviewer: Maybe this?
```suggestion
For each notification, the Push Notification Service receives and forwards to Apple or Google:
```

That lets this section explicitly cover also the question of what Google and Apple see.

>   service from registration; never seen in plaintext by the Zulip server).

Reviewer: This item doesn't seem to fit this description:
```suggestion
For each notification, the Push Notification Service receives and forwards:

```

I think the ways it does appear are adequately covered in the other sections below.

>   ciphertext), forwarded to Apple or Google inside the push payload.

Reviewer: The mention of libsodium and the name of a function within it are I think more detailed than is relevant to a user, even a well-informed technical user. Instead let's name the actual cryptosystem:
```suggestion
- The encrypted notification payload (an XSalsa20+Poly1305
  ciphertext), forwarded to Apple or Google inside the push payload.
```

>   removal) and the size of the encrypted payload.

Reviewer: The size isn't really a separate piece of information the service sees; we already say it sees the encrypted payload itself. I think I'd just cut this mention.

A thorough security analysis would point out that the size is potentially informative to an eavesdropper. If we want to point that out, I'd put it either within the payload item above, or as a separate comment below; grouping it here with a different piece of information feels odd.

Author: Dropping it here.

> over the wire with SSL/TLS.

Reviewer: This belongs in one of the other sections, not this section which is about data that's stored.

Author: Moved to the "what they see" section.

> decrypt.

Reviewer: This can be merged into the section about what we see, because the answers are the same.

The first paragraph is a summary of what's given in more detail there. The second paragraph is a useful point to emphasize which applies to our service too.

malicious Zulip server. Learn about [Zulip's approach to
security](https://zulip.com/security/), and [keep your server
safe](./securing-your-zulip-server.md).

### Limitations

The E2EE protocol has the following known limitations:

- **No forward secrecy by default.** A device's symmetric key persists
  until the mobile app rotates it, and is stored by both the mobile app
  and the Zulip server. An attacker who obtains both a historical capture
  of encrypted push payloads and the device's current key — for example,
  by compromising the mobile device or the Zulip server — can decrypt

>   previously captured notifications.

Reviewer: We can say something better here. It's true the key persists until rotated — but the app does rotate them, generally every 30 days.

And then at the end: that attacker can only decrypt previously captured notifications recent enough that they were encrypted with the key the attacker got.

Author: Edited to:

> - **Incomplete forward secrecy.** A device's symmetric key persists until the
  mobile app rotates it (generally every 30 days), and is stored by both the
  mobile app and the Zulip server. An attacker who obtains both a historical
  capture of encrypted push payloads and the device's current key — for example,
  by compromising the mobile device or the Zulip server — can decrypt the
  notifications captured since the key was last rotated.

Reviewer: LGTM!

- **Metadata side channels.** The Push Notification Service, and
  indirectly Apple and Google, can observe the timing, delivery priority,
  and size of each push notification. These side channels can reveal

>   coarse patterns of activity even when message contents are encrypted.

Reviewer: Cool, this analysis section seems like a good place to mention the size. I don't think there's a need to mention it also in the more just-the-facts sections above.

>   coarse patterns of activity even when message contents are encrypted.

Reviewer: Perhaps this?
```suggestion
  and size of each push notification. These side channels can reveal
  coarse patterns of activity even though message contents are encrypted.
```

- **Lost devices.** The mobile app stores its copy of the per-device
  symmetric key locally, without backup. A user who loses access to a

>   device cannot recover its historical notifications; reinstalling the

Reviewer: Recovering historical notifications isn't really a thing anyway; I don't think this is a limitation.

Author: Dropping this bullet.

  app generates a fresh key and re-registers the device with your Zulip

>   server and the Push Notification Service.

Reviewer: And then I think there isn't really a limitation that remains here.

(We do have a potential user annoyance in that on a new device the user has to log in afresh. But that's independent of notifications. Once they do log in, the new device will get notifications just like the old one did.)

### Cryptography

The E2EE protocol encrypts each notification payload with libsodium's
`crypto_secretbox_easy` (XSalsa20-Poly1305 authenticated encryption) using

> a per-device 32-byte symmetric key and a fresh 192-bit random nonce per

Reviewer: Crypto key sizes are quoted in bits:
```suggestion
a per-device 256-bit symmetric key and a fresh 192-bit random nonce per
```

notification. Initial registration of a device with the Push Notification
Service uses libsodium's `crypto_box_seal` (Curve25519 + XSalsa20-Poly1305)
so that the Zulip server can forward the device's APNs/FCM token to the
bouncer without learning the token itself. A one-byte algorithm prefix on

> each symmetric key reserves space for future cipher migrations.

Reviewer: Shrug, I guess in this section of its own it's fine to have this extra detail.

`/etc/zulip/settings.py` file (i.e., add `# ` at the start of the line), and
[restart your Zulip server](settings.md#changing-server-settings). This approach makes it
easy to start using the service again by uncommenting the same line.

### Legacy protocol

> For mobile clients that do not yet support E2EE, the Zulip uses a legacy

Reviewer: s/the //?

protocol. The central design goal of the legacy Push Notification Service
protocol is to avoid any message content being stored or logged by the service,

> even in error cases.

Reviewer: This is really more about the service than the protocol; in particular, the fact we don't store notification contents is something the protocol itself doesn't at all help with.

    registered to be notified by your Zulip server.
  - User ID numbers generated by your Zulip server, needed to route
    a given notification to the appropriate set of mobile devices.
    These user ID numbers are opaque to the Push Notification
    Service and Kandra Labs.
  - [Basic organization metadata](#uploading-basic-metadata),
    [optional usage statistics](#uploading-usage-statistics), and
    aggregate statistics about how many push notifications are sent by
    each customer.
- The Push Notification Service receives (but does not store) the
  contents of individual mobile push notifications:

  - The numeric message ID generated by your Zulip server.
  - Metadata on the message's sender (name and avatar URL).
  - Metadata on the message's recipient (channel name + ID, topic,
    direct message recipients, etc.).
  - A timestamp.
  - The message's content.

  Zulip 11.0+ has an organization-level setting available to disable
  message content being sent via the push notification bouncer (i.e.,
  message content will be replaced with `New message`), for clients
  that don't support the new end-to-end encrypted notifications

>   protocol. (Prior to Zulip 11.0, this functionality was available via the

Reviewer: There's an open PR to strengthen this setting so that it disables legacy notifications entirely (so in particular the sender and recipient data doesn't get sent either). I guess this can merge first, and then that PR can edit this to describe its own effect.

Reviewer: (OK, and in fact that PR #38433 already does that to the old version of this paragrah. So this comment just comes down to that PR properly resolving the conflict when rebased atop this one.)
